### PR TITLE
Prevent discard of value names for -g3 or above

### PR DESCRIPTION
**Pull request by milizhang**
This change is another part that aims to solve #7883, by adding "-fno-discard-value-names" to clang invocation calls from emcc.py, when debug levels >= 3.

It depends on emscripten-core/emscripten-fastcomp-clang#27.

### CODE REVIEW

File: emcc.py [hunk 1]
```diff
@@ -1565,6 +1565,9 @@ def get_bitcode_file(input_file):
           options.debug_level = 4
           shared.Settings.DEBUG_LEVEL = 4
 
+      if options.debug_level > 3:
+          newargs.append('-fno-discard-value-names')
```
Comment (sbc100) [inline emcc.py]: We probably only want this for ASMJS so I would add "and if not Settings.WASM". 

Also we use two spaced for indentation.

Otherwise lgtm

Comment (milizhang) [reply emcc.py]: Sorry I did not test the upstream LLVM-WASM backend. Luckily seems like the automation also caught it.
Will fix this soon.

File: emcc.py [hunk 2]
```diff
@@ -1565,8 +1565,8 @@ def get_bitcode_file(input_file):
           options.debug_level = 4
           shared.Settings.DEBUG_LEVEL = 4
 
-      if options.debug_level > 3:
-          newargs.append('-fno-discard-value-names')
+      if options.debug_level > 3 and not shared.Settings.WASM_BACKEND:
```
Comment (sbc100) [inline emcc.py]: I think you should check "not shared.Settings.WASM" instead.. as this is an asmjs-only thing AFAICT.

Comment (sbc100) [reply emcc.py]: Perhaps with a comment:  "Unlike other targets, asmjs can preserver the internal llvm names in the final output..."

File: emcc.py (hunk 1 above)
Comment (kripken) [reply emcc.py]: This should work with the LLVM wasm backend too, I think? We backported this option from newer clang, so upstream should have the option.

Comment (kripken) [reply emcc.py]: Oh, is the issue that the wasm backend just can't preserve names anyhow? I see then.

File: emcc.py [hunk 3]
```diff
@@ -1565,6 +1565,10 @@ def get_bitcode_file(input_file):
           options.debug_level = 4
           shared.Settings.DEBUG_LEVEL = 4
 
+      # For asm.js, the generated JavaScript could preserve LLVM value names, which can be useful for debugging.
+      if options.debug_level > 3 and not shared.Settings.WASM:
```
Comment (kripken) [inline emcc.py]: I think this should be `>= 3`?

**Comment by kripken** [comment]
I tagged 1.38.29 now, which will include the clang change in the binary build. Once the builds are done (usually takes a few hours), we can merge that to here, and then tests should pass.

File: emcc.py (hunk 3 above)
Comment (milizhang) [reply emcc.py]: For some reason I thought that I wrote >= when I submitted this.

Fixed.

**Comment by kripken** [comment]
Sadly the 1.38.29 tag failed - but we just landed another change that should fix things. If you merge very latest incoming here now then tests should be green.

**Comment by kripken** [comment]
Thanks @milizhang , code looks good now. I don't see a merge of latest incoming into this PR, though? Please do that so that the tests can pass (it's the only way to use a new build of LLVM+clang with your changes, otherwise the new flag isn't recognized).

**Comment by milizhang** [comment]
Seems like test_sdl2_timer failed on FireFox, but for a reason I do not really understand as the diff does not seem to show anything meaningful. @kripken How do you think about this?

**Comment by kripken** [comment]
That test failure is probably unrelated, it fails randomly now and then and we haven't figured out why yet. So I think everything looks good here! Except I see we forgot to ask you to add yourself to AUTHORS earlier, sorry about that - please do so before merge.

**Comment by kripken** [comment]
Perfect, thanks @milizhang!

**Pull request merged**